# ------------------------------------------------------------------------------------------------ #
# Syri results
#
# This script takes the '.out' file generated by Syri and generates a simple summary file from the
# results with the following:
#   - Count of each type of feature
#   - Cumulative sum of the length of each feature in the reference and query genomes
#   - Min/median/max and 1st/3rd quartile ranges for reference and query genomes from lengths
#
# This script doesn't consider SNPs.

# ------------------------------------------------------------------------------------------------ #
# Libraries
suppressPackageStartupMessages({
  library(tidyverse)
  library(here)
})

# ------------------------------------------------------------------------------------------------ #
# Functions

# Read the syri output file as a tibble
read_syri <- function(path, name) {
  read_tsv(
    file = path,
    col_names = c(
      'chr_ref', 'start_ref', 'end_ref', 'seq_ref',
      'seq_qry', 'chr_qry', 'start_qry', 'end_qry',
      'uniqueID', 'parentID', 'annotation_type', 'copy_status'
    ),
    col_types = cols(),
    col_select = c(-seq_ref, -seq_qry)
  ) |>
    mutate(
      across(
        starts_with(c('start_', 'end_')),
        as.numeric
      ),
      comparison = name
    ) |>
    select(comparison, everything())
}

# Calculate the widths of features (ignoring SNPs as their width is one)
width_syri <- function(df) {
  df |>
    filter(annotation_type != 'SNP') |>
    mutate(
      width_ref = abs(end_ref - start_ref),
      width_qry = abs(end_qry - start_qry)
    ) |>
    select(comparison, uniqueID, parentID, annotation_type, width_ref, width_qry)
}

# Generate the summary statistics for count and total length of features
summarise_syri <- function(df) {
  lvls <- c(
    'SYN','INV','TRANS','INVTR','DUP','INVDP','NOTAL',
    'CPG','HDR','INS','SYNAL','INVAL','TRANSAL',
    'INVTRAL','DUPAL','INVDPAL','SNP','CPL','TDM','DEL'
  )
  df |>
    group_by(comparison, annotation_type) |>
    mutate(annotation_type = factor(annotation_type, levels = lvls)) |>
    summarise(
      count = n(),
      ref = sum(width_ref, na.rm = TRUE),
      qry = sum(width_qry, na.rm = TRUE)
    ) |>
    arrange(annotation_type) |>
    pivot_longer(
      names_to = 'genome',
      values_to = 'sum_len',
      ref:qry
    )
}

# Generate the length summary statistics for each feature
iqr_width_syri <- function(df) {
  val <- df |> pull(comparison) |> unique()
  df |>
    group_by(annotation_type) |>
    (\(x) split(x, x$annotation_type))() |>
    map(\(x) {
      ref <- x |>
        pull(width_ref) |>
        summary() |>
        unclass() |>
        (\(y) tibble(names = names(y), values = y))() |>
        pivot_wider(names_from = names, values_from = values) |>
        mutate(genome = 'ref')

      qry <- x |>
        pull(width_qry) |>
        summary() |>
        unclass() |>
        (\(y) tibble(names = names(y), values = y))() |>
        pivot_wider(names_from = names, values_from = values) |>
        mutate(genome = 'qry')

      bind_rows(ref, qry)
    }) |>
    list_rbind(names_to = 'annotation_type') |>
    mutate(comparison = val) |>
    select(comparison, everything(), -`NA's`)
}

# Run all of the above functions in one step
syri_build <- function(path, name) {
  df <- read_syri(path, name)
  df.width <- width_syri(df)
  df.summary <- summarise_syri(df.width)
  df.iqr <- iqr_width_syri(df.width)

  df.final <- df.summary |>
    left_join(df.iqr) |>
    select(comparison, genome, annotation_type, count, everything())

  rm(df, df.width, df.summary, df.iqr)
  invisible(gc())
  return(df.final)
}

# ------------------------------------------------------------------------------------------------ #
# Process all output files
syri.results <- fs::dir_ls(
  path = here('synteny','syri','results', 'syri'),
  glob = '*.out'
) |>
  as.character() |>
  (\(x) x[str_detect(x, 'th-elegans', negate = TRUE)])() |>
  (\(x) set_names(x, sub('.syri.*', '', basename(x))))() |>
  imap(\(x,y) {syri_build( x, y)}) |>
  list_rbind() |>
  select(-`1st Qu.`, -`3rd Qu.`) |>
  rename(
    `No. Features` = count,
    `Total length` = sum_len,
    Min = `Min.`,
    Max = `Max.`,
    Feature = annotation_type
  )

# ------------------------------------------------------------------------------------------------ #
# Make a summary table
syri.results |>
  mutate(
    `Count (length bp)` = paste0(`No. Features`, ' (', scales::comma(`Total length`), ')'),
    comparison = case_when(
      comparison == 'ornatus_major' ~ 'H. ornatus - H. major',
      comparison == 'major_curtus-AG' ~ 'H. major - H. curtus (AG)',
      comparison == 'curtus-AG_curtus' ~ 'H. curtus (AG) - H. curtus',
      comparison == 'curtus_cyanocinctus' ~ 'H. curtus - H. cyanocinctus'
    ),
    comparison = factor(comparison, levels = c(
      'H. ornatus - H. major', 'H. major - H. curtus (AG)',
      'H. curtus (AG) - H. curtus', 'H. curtus - H. cyanocinctus')
    ),
    genome = ifelse(genome == 'ref', 'Reference', 'Query'),
    genome = factor(genome, levels = c('Reference', 'Query')),
    Description = case_when(
      Feature == 'SYN' ~ 'Syntenic region',
      Feature == 'INV' ~ 'Inverted region',
      Feature == 'TRANS' ~ 'Translocated region',
      Feature == 'INVTR' ~ 'Inverted transolcated region',
      Feature == 'DUP' ~ 'Duplicated region',
      Feature == 'INVDP' ~ 'Inverted duplicated region',
      Feature == 'NOTAL' ~ 'Unaligned region',
      Feature == 'CPG' ~ 'Copy gain (query)',
      Feature == 'HDR' ~ 'Highly diverged region',
      Feature == 'INS' ~ 'Insertion (query)',
      Feature == 'SYNAL' ~ 'Alignment in syntenic region',
      Feature == 'INVAL' ~ 'Alignment in inverted region',
      Feature == 'TRANSAL' ~ 'Alignment in translocated region',
      Feature == 'INVTRAL' ~ 'Alignment in inverted translocated region',
      Feature == 'DUPAL' ~ 'Alignment in duplicated region',
      Feature == 'INVDPAL' ~ 'Alignment in inverted duplicated region',
      Feature == 'SNP' ~ 'Single nucleotide polymorphism',
      Feature == 'CPL' ~ 'Copy loss (query)',
      Feature == 'TDM' ~ 'Tandem repeat',
      Feature == 'DEL' ~ 'Deletion (query)'
    ),
    Feature = factor(Feature, levels = c(
      'SYN','INV','TRANS','INVTR','DUP','INVDP','NOTAL',
      'CPG','HDR','INS','SYNAL','INVAL','TRANSAL',
      'INVTRAL','DUPAL','INVDPAL','SNP','CPL','TDM','DEL'
    ))
  ) |>
  select(Feature, Description, comparison, Genome = genome, `Count (length bp)`) |>
  arrange(Feature, Genome, comparison) |>
  pivot_wider(names_from = comparison, values_from = `Count (length bp)`) |>
  write_csv(
    file = here('tables','table-x-syri-summary.csv'),
    col_names = TRUE
  )
